gradle/render-javadoc.gradle

```diff
@@ -15,93 +15,105 @@
  * limitations under the License.
  */
 
-// generate javadocs by using Ant javadoc task
+// generate javadocs by calling javadoc tool
+// see https://docs.oracle.com/en/java/javase/11/tools/javadoc.html
+
+// utility function to convert project path to document output dir
+// e.g.: ':lucene:analysis:common' => 'analysis/common'
+def pathToDocdir = { path -> path.split(':').drop(2).join('/') }
 
 allprojects {
   plugins.withType(JavaPlugin) {
-    ext {
-      javadocRoot = project.path.startsWith(':lucene') ? project(':lucene').file("build/docs") : project(':solr').file("build/docs")
-      javadocDestDir = "${javadocRoot}/${project.name}"
-    }
-
     task renderJavadoc {
-      description "Generates Javadoc API documentation for the main source code. This invokes Ant Javadoc Task."
+      description "Generates Javadoc API documentation for the main source code. This directly invokes javadoc tool."
       group "documentation"
 
       ext {
-        linksource = "no"
+        linksource = false
         linkJUnit = false
-        linkHref = []
+        linkLuceneProjects = []
+        linkSorlProjects = []
       }
 
       dependsOn sourceSets.main.compileClasspath
 
       inputs.files { sourceSets.main.java.asFileTree }
-      outputs.dir project.javadocRoot
+      outputs.dir project.javadoc.destinationDir
 
       def libName = project.path.startsWith(":lucene") ? "Lucene" : "Solr"
       def title = "${libName} ${project.version} ${project.name} API".toString()
 
+      // absolute urls for "-linkoffline" option
+      def javaSEDocUrl = "https://docs.oracle.com/en/java/javase/11/docs/api/"
+      def junitDocUrl = "https://junit.org/junit4/javadoc/4.12/"
+      def luceneDocUrl = "https://lucene.apache.org/core/${project.version.replace(".", "_")}".toString()
+      def solrDocUrl = "https://lucene.apache.org/solr/${project.version.replace(".", "_")}".toString()
+
+      def javadocCmd = org.gradle.internal.jvm.Jvm.current().getJavadocExecutable()
```

Comment:
Not sure if it's using Gradle's "internal" JVM or the "project's" JVM. We should be using the javadoc tool that is aligned with what is used for the compiler and execution (tests). I think you got it right but maybe worth double checking.

https://docs.gradle.org/current/userguide/build_environment.html
(what happens if you set org.gradle.java.home=... to an older JVM, for example).

Reply:
I set the default java home to JDK14 (that is used to launch gradle) and "org.gradle.java.home" to JDK11.
```
$ echo $JAVA_HOME
/Library/Java/JavaVirtualMachines/adoptopenjdk-14.jdk/Contents/Home
$ java -version
openjdk version "14" 2020-03-17
$ cat gradle.properties
org.gradle.java.home=/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home
```

I did just print debug for `org.gradle.internal.jvm.Jvm.current().getJavadocExecutable()`, that said the `JVM.current()` points the same jvm to org.gradle.java.home property.
```
/Library/Java/JavaVirtualMachines/adoptopenjdk-11.jdk/Contents/Home/bin/javadoc
```

Is this expected behaviour (or is there another configuration point)?

Reply:
I'm not sure, to be honest. In that setting, which JVM is used for running tests? My assumption was that it'd be the first one one from path (or JAVA_HOME) but maybe I'm wrong.

Reply:
> In that setting, which JVM is used for running tests?

Gradle Test task has a customizable property `executable`, the documentation says "The name of the executable to use. Default with java plugin: java command for the current JVM." If my understanding is correct, it is the runtime on which tests are run.
https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.Test.html

I tried to print test's executable and javadoc's executable by adding doLast{} clause on the default test task (gradle/default-tests.gradle).
```
      doLast {
        println("Test executable\t " + executable)
        println("Javadoc executable\t" + org.gradle.internal.jvm.Jvm.current().getJavadocExecutable())
      }
```

I set 
`JAVA_HOME=/usr/local/java/adoptopenjdk/jdk-14+36/` and 
`org.gradle.java.home=/usr/local/java/adoptopenjdk/jdk-11.0.3+7/`
then, ran the tests. It said both `executable` point the gradle's java home. (Here we don't change the executable for test, so the default "current" JVM should be used.)

```
Test executable        /usr/local/java/adoptopenjdk/jdk-11.0.3+7/bin/java
Javadoc executable      /usr/local/java/adoptopenjdk/jdk-11.0.3+7/bin/javadoc
```

Looks consistent to me, though i could miss something here... what do you think @dweiss ?

Reply:
I really don't know the details here. Thank you for investigating! I know it is possible to separate gradle's JVM and the JVM used for everything else but I never did this. ES folks do have this separation implemented but I only skipped through that code a while ago.

https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java

Reply:
I did a little more experiments with `org.gradle.internal.jvm.Jvm.current()`, which is used both on compilation and test execution, the search path is
1. org.gradle.java.home
2. $JAVA_HOME
3. user's default java (on $PATH)

It's consistent with their documentation.

Elasticsearch's custom build plugin takes completely different search strategy from gradle's.
1. "compiler.java" system property
2. $JAVA_HOME
3. org.gradle.java.home
4. user's default java (on $PATH)

(I didn't run it but just interpreted this method https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/java/org/elasticsearch/gradle/info/GlobalBuildInfoPlugin.java#L209)

Reply:
I am going to merge it to master branch since I think I understand what I did here with `org.gradle.internal.jvm.Jvm.current()`.

Reply:
Thanks for looking into this, Tomoko. We may have to do something similar to what ES does since we want to be able to run javac, javadocs and tests against new JVMs (which gradle itself may not support yet). It's a different issue though and it can certainly wait.

Reply:
I just merged it to the master. 

> We may have to do something similar to what ES does since we want to be able to run javac, javadocs and tests against new JVMs (which gradle itself may not support yet).

Should we open an issue for that, or can it be delayed?

Reply:
I've opened https://issues.apache.org/jira/browse/LUCENE-9312 and will perhaps take a look at it this Eastern -- feel free to join! It is important and your investigation helps.